### DISCUSSION TITLE
Componet Still Renders Despite $this->skipRender();

### DISCUSSION THREAD

**Discussion by mrl22** [livewire/livewire#5353] [1. Help]
Hello

I have been working on a JavaScript class that creates draggable process maps, nodes are dragged onto a board, and their existence and positioning are updated through javascript Livewire.emit() events in my class. My class works directly from the DOM and does not store any data in JavaScript objects or arrays, so as Livewire changes, my class can read new data without re-initialisation.

[https://www.dropbox.com/s/du08ctzq83zan05/Screenshot%202022-11-09%20at%2002.57.58%20pm.png?dl=0](https://www.dropbox.com/s/du08ctzq83zan05/Screenshot%202022-11-09%20at%2002.57.58%20pm.png?dl=0)

In the PHP method, I am saving the position of a node, but Livewire is re-rendering the view, which I do not want it to do.

1. My first step was to wrap my whole JavaScript class in a wire:ignore, which works perfectly with the exception that I cannot (or do not know how to) force a re-render on a wire:ignore element should I need to.

2. My second approach was to add `$this->skipRender();` to my methods which I thought would not re-render the contents of the component at all, but when dragging a node, it re-renders with the old data with the node in the old location. If I refresh the page, the node is in the new location.

3. My third approach was to add `exit;` to my methods, this works perfectly, but I am left with a JavaScript error in the console, which I assume is because Livewire did not get a 200 successful response.

Any advice or things to try would be appreciated.

Javascript Hook
```
window.flow.on('afterDragNodeEnd', (flow) => {
    // Javascript hook for when a node has been dropped into a new position
    let node = flow.dragElement;
    Livewire.emit(
        'moveNode',
        node.getAttribute('data-key'),
        parseInt(node.style.left, 10),
        parseInt(node.style.top, 10)
    );
});
```

PHP Method
```
public function moveNode($uuid, $x, $y)
{

    Route::whereUuid($uuid)
        ->update([
            'x' => $x,
            'y' => $y
        ]);
    $this->skipRender();
}
```

**Comment by mrl22** [comment]
Ok, on further investigation it seems that this line is the issue https://github.com/livewire/livewire/blob/63ad79f9bbad51028e1aaeca774f2f5e22f6578d/js/component/index.js#L293-L295

Despite, `$this->skipRender()` being set and `response.effects.html` = `null` in JavaScript, the DOM is still updated but with the old data...

Maybe I am asking too much of Livewire and should switch to using Laravel API?

**Reply by joshhanley** [reply]
@mrl22 yeah Livewire still does a re-render on each request, all `$this->skipRender()` does is stops the blade engine running render on the view. If I remember rightly, it's to ensure all events and actions happen property when the request finishes does due to it being driven by the html.

If you are using a javascript front end, then I think personally I wouldn't bother using Livewire in that instance and just use an API or InertiaJS for that page.

Hope this helps!

**Reply by joshhanley** [reply]
Thinking about it some more, another way you could do it, is use javascript to optimistically update the node positions in the UI, which sends their new position back to Livewire, then allow Livewire to actually render and ensure it renders the nodes in the new place.

You can see Caleb do that in one of these screencasts (I think) 
https://laravel-livewire.com/screencasts/s8-animate-move
https://laravel-livewire.com/screencasts/s8-dragging-list

Hope this helps

**Comment by Martleby** [comment]
This is a bit hacky.... but with your solution 1 (wire:ignore) you could actually just remove that attribute from the element as and when needed and then put it back again once done.

Cheers,
Martin.

**Comment by mrl22** [comment]
Thank you @joshhanley, appreciate the advice. For now, I adopted the approach you mentioned, making the return from livewire mimic everything that my javascript class is doing.

@Martleby I did play around with removing and adding the wire:ignore but I couldn't get it to work right. (If you see my demo below) the only time I needed it to overwrite was for a board change as you see at the top.

I have set up a demo for you to play if you like. There are bugs and it is not finished. It is a proof of concept for a project.

https://processiq.devfwd.com/

Note that if multiple users are playing with it, it may have strange effects.

There are shortcut keys such as delete/backspace to remove elements. Zoom does not work yet. The board can be dragged and is infinite in size. The board is updated in real time, so if you refresh it should be identical.

**Reply by joshhanley** [reply]
@mrl22 no worries! 🙂
